openmdao/docs/feature_reference/building_blocks/components/file_wrap.rst

```diff
@@ -0,0 +1,820 @@
+define a component that calls an external program to perform it's computation,
+passing input and output values via files.
+
+The input and output files were very simple in that basic example, containing only
+the values of interest.  In the general case however, you will probably need to
+generate an input file with a specific format of rows and fields and to parse a
+similarly formatted output file to get the output values. To facilitate working
+with these more complex input and output files, OpenMDAO provides a couple of utility
+classes: `InputFileGenerator` and `FileParser`.
+
+
+Generating the Input File
+-------------------------
+
+You can generate an input file for an external application in a few different ways.
+One way is to write the file completely from scratch using the new values that are
+contained in the component's variables. Not much can be done to aid with this task, as
+it requires knowledge of the file format and can be completed using Python's standard
+formatted output.
+
+A second way to generate an input file is by templating. A *template* file is
+a sample input file which can be processed by a templating engine to insert
+new values in the appropriate locations. Often the template file is a valid
+input file before being processed, although other times it contains directives
+or conditional logic to guide the generation. Obviously this method works well
+for cases where only a small number of the possible variables and settings are
+being manipulated.
+
+OpenMDAO includes a basic templating capability that allows a template file to
+be read, fields to be replaced with new values, and an input file to be
+generated so that the external application can read it. Suppose you have an
+input file that contains some integer, floating point, and string inputs:
+
+::
+
+    INPUT
+    1 2 3
+    INPUT
+    10.1 20.2 30.3
+    A B C
+
+This is a valid input file for your application, and it can also be used as a
+template file. The templating object is called `InputFileGenerator`, and it
+includes methods that can replace specific fields as measured by their row
+and field numbers.
+
+To use the InputFileGenerator, first instantiate it and give it the name of
+the template file and the name of the output file that you want to produce. (Note
+that this code must be placed in the ``compute`` method of your component *before*
+the external code is run.) The code will generally look like this:
+
+::
+
+    from openmdao.util.file_wrap import InputFileGenerator
+
+    parser = InputFileGenerator()
+    parser.set_template_file('mytemplate.txt')
+    parser.set_generated_file('myinput.txt')
+
+    # (Call functions to poke new values here)
+
+    parser.generate()
+
+When the template file is set, it is read into memory so that all subsequent
+replacements are done without writing the intermediate file to the disk. Once
+all replacements have been made, the ``generate`` method is called to create the
+input file.
+
```

Comment:
Are we testing anything that uses the "testcode" directive? I know we used to test this stuff automatically, but I haven't even been regularly running a `make doctest` since we decided to use embed_test for everything. In fact, running it on master right now, I see 3 failures. Need to either switch this over to use the embed_test directive, or get this testing included as a step in the CI.

Reply:
@Kenneth-T-Moore We could bring back doctest, but I find no evidence that we are currently using it anywhere.

Reply:
switched to using `embed-test` per suggestion